Shoko.Server/API/v3/Controllers/PlexController.cs

```diff
@@ -0,0 +1,203 @@
+[ApiV3]
+[Authorize]
+public class PlexController : BaseController
+{
+    private const string AlreadyAuthenticated = "User has already authenticated with plex yet. Invalidate the current token before re-authenticating.";
+
+    private const string NotAuthenticated = "User has not authenticated with plex yet. Authenticate first before retrying.";
+
+    private const string NoSelectedServer = "A server has not been selected yet. Select a server first.";
+
+    private const string InvalidServerSelection = "Invalid server selection.";
+
+    private const string InvalidLibrarySelection = "Selected libraries are invalid.";
+
+    /// <summary>
+    /// Get an OAuth2 authenticate url to authenticate the current user.
+    /// </summary>
+    /// <returns></returns>
+    [HttpGet("Auth/Authenticate")]
+    public ActionResult<string> GetOAuthRequestUrl([FromQuery] bool force = false)
+    {
+        var helper = GetHelperForUser();
+
+        if (helper.IsAuthenticated && !force)
+            return BadRequest(AlreadyAuthenticated);
+
+        return helper.LoginUrl;
+    }
+
+    /// <summary>
+    /// Check if the current user is currently authenticated against the plex api.
+    /// </summary>
+    /// <returns></returns>
+    [HttpGet("Auth/IsAuthenticated")]
+    public ActionResult<bool> IsAuthenticated()
+    {
+        var helper = GetHelperForUser();
+
+        return helper.IsAuthenticated;
+    }
+
+    /// <summary>
+    /// Invalidate and remove the current plex authentication token.
+    /// </summary>
+    /// <returns></returns>
+    [HttpPost("Auth/InvalidateToken")]
+    public ActionResult<bool> InvalidateToken()
+    {
+        var helper = GetHelperForUser();
+
+        if (!helper.IsAuthenticated)
+            return false;
+
+        helper.InvalidateToken();
+
+        return true;
+    }
+
+    /// <summary>
+    /// Show all available server for the current user.
+    /// </summary>
+    /// <param name="userId"></param>
+    /// <returns></returns>
+    [HttpGet("Linking/AvailableServers")]
+    public ActionResult<List<PlexServer>> AvailableServers(int userId)
+    {
+        var helper = GetHelperForUser();
+
+        if (!helper.IsAuthenticated)
+            return BadRequest(NotAuthenticated);
+
+        var currentServerID = helper.ServerCache?.ClientIdentifier ?? "";
+        return helper.GetPlexServers()
+            .Select(device => new PlexServer(device, device.ClientIdentifier == currentServerID))
+            .ToList();
+    }
+
+    /// <summary>
+    /// Get the active server.
+    /// </summary>
+    /// <returns></returns>
+    [HttpGet("Linking/Server")]
+    public ActionResult<PlexServer> CurrentDevice()
+    {
+        var helper = GetHelperForUser();
+
+        if (!helper.IsAuthenticated)
+            return BadRequest(NotAuthenticated);
+
+        var currentServer = helper.ServerCache;
+        if (currentServer == null)
+            return BadRequest(NoSelectedServer);
```

Comment:
Should an "unlinked" be a bad request? I would expect this to be a http 204

Reply:
That's the difference between loudly complaining (this 400) and silently failing (204). Personally I want it to loudly complain, but silently failing (and returning an empty) body is also fine. Will update it to use 204 instead.

Reply:
Your still returning a 200 series response, aka a successful request, 400 indicates something was wrongly requested on the client & using 204 or “No content” is still distinct enough if you want to determine based off the return code